packages/design-system/src/components/optimized-image/optimized-image.ts

```diff
@@ -0,0 +1,211 @@
+   * OptimizedImage({
+   *   src: 'path/to/some/image.webp',
+   *   srcset: '100w, 200w'
+   * })
+   * // Result: `<img src="path/to/some/image.webp" srcset="path/to/some/image-100w.webp 100w, path/to/some/image-200w.webp 200w">`
+   */
+  srcset?: string;
+
+  /**
+   * Specifies sizes attribute for the image.
+   */
+  sizes?: string;
+
+  /**
+   * A function used to rewrite `srcset` attributes or create placeholder images.
+   *
+   * If this function is used to create a placeholder image, the `isPlaceholder`
+   * field in the `ImageLoaderConfig` object will be set to `true`
+   *
+   */
+  loader?: Loader;
+}
+
+const createPlaceholder = (img: BaseComponent<HTMLImageElement>, placeholder: string, blur?: number) => {
+  const styles: Partial<CSSStyleDeclaration> = {
+    filter: `blur(${blur ?? DefaultBlurAmount}px)`,
+    backgroundImage: `url(${placeholder})`,
+    backgroundPosition: '50%, 50%',
+    backgroundRepeat: 'no-repeat',
+    backgroundSize: 'cover',
+    overflow: 'hidden',
+  };
+
+  img.stylize(styles);
+
+  return () => {
+    img.unstylize(...keys(styles));
+  };
+};
+
+const fill = (img: BaseComponent<HTMLImageElement>) => {
+  img.stylize({
+    width: '100%',
+    height: '100%',
+    inset: '0',
+    position: 'absolute',
+  });
+};
+
+/**
+ * HTMLImageElement that has fetchpriority because typescript does not include
+ * fetchpriority in standart HTMLImageElement
+ */
+export type OptimizedImageElement = HTMLImageElement & { fetchpriority: Priority };
+
+/** @internal */
+export const lazyAssertNoDistortion = (img: BaseComponent<OptimizedImageElement>, width: number, height: number) => {
+  img.once('load', () =>
+    import('./utils/distortion.js').then((module) => module.assertNoDistortion(img, width, height)),
+  );
+};
+
+const getPlaceholderUrl = (props: OptimizedImageProps) => {
+  return props.placeholder === true
+    ? validateFn(props.loader!, { src: props.src, isPlaceholder: true, widthAsNumber: props.width })
+    : props.placeholder!;
+};
+
+/**
+ * Optmized image component which enforces best practices for loading images.
+ * Warns if image is distroted and shows how to fix it.
+ * Make sure to add `<link rel="preload" as="image" href="<image src here">` tag in head if the image is high priority.
+ * @see {@link OptimizedImageProps}
+ * @returns new `OptimizedImage`
+ *
+ * @example
+ * OptimizedImage({
+ *   src: 'path/to/some/kitty.webp',
+ *   width: 500,
+ *   height: 500,
+ *   alt: 'cute kitty image'
+ * })
+ */
+export const OptimizedImage = (props: OptimizedImageProps) => {
+  const {
+    src,
+    laziness = 'lazy',
+    width,
+    height,
+    alt,
+    blur,
+    fill: isFill,
+    placeholder,
+    srcset,
+    sizes,
+  } = validateProps(props);
+
+  const img = new BaseComponent<OptimizedImageElement>({
+    src,
+    width,
+    height,
+    alt,
+    sizes,
+    ...createImageProps(laziness),
+  });
+
+  if (srcset) {
+    img.node.srcset = generateSrcset(src, srcset, width!);
```

Comment:
Please, add correct typings for your `validateProps`, for example, where if `srcset` is available => width isn't nullable. Use discriminated unions feature.